### ISSUE TITLE
Evaluating a derivative elementwise

### ISSUE THREAD

**Issue by c-f-h** [HIPS/autograd#250]
Maybe I'm using the library wrong, but I wonder if what I'm observing is a bug or intended. I'm using the latest Github master version.

Say I have a scalar function of two variables x and y, but I want to retain the option to pass in numpy arrays and get the values/derivatives elementwise. I tried the following:

```
import numpy as np
from autograd import grad

f = lambda x,y: (x+y)**2
fgrad = grad(f, 0)

x = np.linspace(0, 1, 11)

fgrad(x, 0.0)
>>> array([ 0. ,  0.2,  0.4,  0.6,  0.8,  1. ,  1.2,  1.4,  1.6,  1.8,  2. ])
```

Looks correct, it computes ``2*x`` as it should. But strange things happen if I want to evaluate the derivative on several points lying along the y axis:

```
fgrad(0.0, x)
>>> 11.000000000000002
```

I would expect to get back an array of the x-derivatives at the points (0,0), (0,0.1), (0,0.2), and so on. Instead I get a single value, and I don't even know where this value 11 comes from -- it seems to be related to the length of the array x?

**Comment by j-towns** [comment]
Yeah I can see why this is confusing. The `grad` function has only supported this kind of broadcasting since a [recent commit](https://github.com/HIPS/autograd/commit/fabac3f896de86f255fa7dbf6ede51192a24baa3), and I think we may not have properly documented it yet.

When you do
```python
fgrad(0.0, x)
```
you're getting the _sum_ of all the derivatives of the elements of f(0.0, x) w.r.t. its first argument: 

2 * 0. + 2 * 0.2 + ... + 2 * 2. = 11.

This behaviour ought to be explained in the tutorial.

**Comment by j-towns** [comment]
A rule of thumb is that the result of `grad(*args, argnum)` will always have the same shape as `args[argnum]`, i.e. it will always have the same shape as the element that you've differentiated w.r.t.

When you do
```python
fgrad(0.0, x)
```
you're differentiating w.r.t. `0.0`, so the returned value will be a float. When you call
```python
fgrad(x, 0.0)
```
you're differentiating w.r.t. `x`, which has shape `(11,)`, thus the returned array has that shape too.

**Comment by mattjj** [comment]
+1 for the rule that the returned grad always has the same shape as the value passed in for the argument with respect to which you're differentiating.

(By the way, your example code has `import numpy as np` at the top, but for autograd to work correctly with numpy functions you need to use `import autograd.numpy as np` instead. That's not an issue for your example code, though.)

In cases like your example code, one solution would be to broadcast manually before you evaluate the function, rather than leaving the broadcasting to the `+` inside the function:

```python
import autograd.numpy as np
from autograd import grad

f = lambda x,y: (x+y)**2
fgrad = grad(f, 0)

x = np.linspace(0, 1, 11)
y = 0.0

x, y = np.broadcast_arrays(x, y)
```

Then `fgrad(x, y)` has the same value as `fgrad(y, x)`.

**Comment by dougalm** [comment]
Yes, the new semantics of `grad` can be a bit confusing. We've always had precisely the same semantics in `elementwise_grad` so the confusion itself isn't new. I'll try to explain the behavior here.

Originally, `grad` was only for scalar-output functions. We recently generalized it to accept array-output functions. The new definition of grad is actually very simple: `grad(f)` applies the transpose of the Jacobian of `f` to a vector of ones. Under some circumstances we can interpret this as a form of broadcasting (but see terminology note below). With a one-argument function `f` we can consider four cases:
  1. `f(x)` is a scalar. In this case `grad(f)(x)` gives the usual gradient.
  2. `f(x)` is an array with the same shape as `x`, and the Jacobian of `f` is diagonal. In this case, left-multiplying the Jacobian by a vector of ones pulls out the diagonal as a vector. So each `grad(f)(x)[i]` represents the derivative of `f(x)[i]` with respect to `x[i]`, which is analogous to numpy's usual automatic mapping by which scalar functions are applied elementwise to arrays.
  3. `f(x)` is an array with each dimension either equal to 1 or equal to the corresponding dimension of x (e.g. `x.shape == (2,3,4)` and `f(x).shape == (2,1,4)`) and the elements of the Jacobian corresponding to different input and output indices are all zero, with the exception of indices for the singleton output dimensions. In this case we end up with automatic mapping over non-singleton dimensions of `f(x)` and regular gradients over singleton dimensions of `f(x)`. Note that case 3 (which actually subsumes 1 and 2) only represents a subset of the cases for which `x` and `f(x)` are numpy-broadcastable. Specifically, `x` can't have singleton dimensions that don't correspond to singletons in `f(x)`. 
  4. In all other cases (such as yours) `grad(f)(x)` doesn't have much interpretation besides "the vector-Jacobian product of a vector of ones" or, equivalently, "the gradient of `lambda x: sum(f(x))`". We should consider having this case raise an error rather than returning a head-scratching result.

@mattjj and @j-towns, we often describe the new behavior of `grad` as 'broadcasting'. Broadcasting (whether numpy or radio) refers to a one-to-many fanout. I'd rather describe the new behavior of `grad` as "automatic mapping". I wonder if changing our terminology might help avoid confusion. Thoughts?

**Comment by mattjj** [comment]
I agree that it's not really "broadcasting" and so we should probably adjust our terminology to avoid confusion. Maybe the right term is ["vectorizing"](https://docs.scipy.org/doc/numpy-1.12.0/reference/generated/numpy.vectorize.html).

Btw, at first I mis-parsed your term "automatic mapping". I think it would be clearer to say "automatic map-ing" or "automatic `map`ing". Or maybe I'm crazy.

Here's a first attempt at drafting how we might summarize it in our documentation (please improve this, or counter with something else!):

For scalar-output functions `f`, `grad(f)` computes the gradient you expect. For vector-output functions `f`, if we have that `f(x)[i] == f(x[i])`, as in the "vectorized" or "map" case where scalar-input numpy functions like `np.tanh` are called on an ndarray input, then we have `grad(f)(x)[i] == grad(f)(x[i])`. (For more general vector-output functions, `grad(f)` evaluates the Jacobian-adjoint operator applied to an all-ones vector, so that `grad(f)(x) == np.tensordot(jacobian(f)(x).H, np.ones_like(f(x)), f(x).ndim)`.)

(That last bit is probably wrong, especially for complex numbers, but it gets the idea across!

**Comment by mattjj** [comment]
By the way, a more abstract thing to say is that if `f` commutes with the coordinate functions, then so does `grad(f)`.

**Comment by c-f-h** [comment]
Thanks all for the very detailed explanations.

Getting the sum of the derivatives was indeed very unexpected for me. Thanks to your explanations, I roughly understand why it happens, but is this result only an artefact of the current implementation or is this ever the correct, desired output?

Is there any way to get the behavior that I wanted, save for manual broadcasting, perhaps by adding a new convenience wrapper? From what I've seen, you guys are mostly interested in ML applications where this is probably not important, but in, say, finite element code, where people work a lot with scalar and vector fields, it would be highly convenient to have a builtin way of getting a derivative which exactly retains the broadcasting behavior of the original function `f`.

**Comment by mattjj** [comment]
It's the desired output. The alternative is just to raise an error when you try to use `grad` with a non-scalar-valued function (which is what we did up until a couple weeks ago when we started trying out this change).

Would the new convenience wrapper just call `np.broadcast_arrays` on some or all inputs before passing them to the function?

**Comment by c-f-h** [comment]
I assume so. The mental model to have is that there is a use case where `f` is a scalar field, i.e., a scalar function which depends on several spatial variables, and to allow broadcasting for all of these.

Would it make sense to have something like this in autograd?

An additional convenience would be to provide a function which returns the gradient of a scalar field; this would be a vector field containing all the partial derivatives. This may be too specialized for inclusion in the library, though.

**Comment by dougalm** [comment]
`f` being a scalar field is actually the main intended use-case for Autograd. Though we're usually thinking of a field in many-dimensional parameter space rather than three-dimensional physical space.

Is the issue that you want to evaluate the gradient at every point of a grid? The vectorization behavior of `grad` should be able to handle it that too. Maybe you could provide a concrete example?

@mattjj - I agree, let's use "vectorize" instead of "automatic map". It's simpler and doesn't evoke cartography. I like your proposed explanation for the docs. It doesn't quite capture case 3 of my previous comment, where `f(x)` has some singleton dimensions and some non-singleton dimensions. Your commutation description captures it though, if we replace "coordinate functions" for appropriately defined "slice functions". For example, if `f` is a function that maps an (L, M, N) array to a (L, 1, N) array, then the relevant set of slice functions is `lambda A: A[i, :, k]`. But a sophisticated user can figure out the behavior from the "vjp applied to ones" definition anyway so it's probably not worth going into the mixed-singleton-non-singleton case.

**Comment by c-f-h** [comment]
Yes, I want to evaluate the gradient over an entire grid. One concrete example is in my original post.

**Comment by dougalm** [comment]
> An additional convenience would be to provide a function which returns the gradient of a scalar field; this would be a vector field containing all the partial derivatives. This may be too specialized for inclusion in the library, though.

> Yes, I want to evaluate the gradient over an entire grid. 

I think this is exactly what you get with `grad`'s vectorizing behavior, provided you set up the inputs correctly. For example, consider the electric potential due to a pair of uniformly charged, nonconducting, hollow spheres. `grad` gives you the electric field in a single step.

```python
import autograd.numpy as np
from autograd import grad

def sphere_potential(xy):
    R = 0.5
    r = np.sqrt(np.sum(xy**2, axis=-1))
    return np.minimum(1/r, 1/R)

def potential(xy):
    sphere_pos = np.array([0., 1.])
    return (  sphere_potential(xy - sphere_pos)
            - sphere_potential(xy + sphere_pos))

X, Y = np.meshgrid(*[np.linspace(-2, 2, 50)]*2)
xy_grid = np.stack([X, Y], axis=-1)

potential_grid = potential(xy_grid)
E_field_grid = -grad(potential)(xy_grid)  # <--- evaluating the electric field

import matplotlib.pyplot as plt
plt.figure()
plt.pcolormesh(X, Y, potential_grid)
plt.quiver(X, Y, E_field_grid[..., 0], E_field_grid[..., 1])
```

![field](https://user-images.githubusercontent.com/8833761/27840021-af6ca2e8-60c3-11e7-8b37-20eb322c4623.png)

**Comment by c-f-h** [comment]
Ah, so it's possible to get the entire gradient at once, that is neat. (Is this in the documentation anywhere? I wasn't aware that this works.)

I still need to keep x and y as separate arguments though, while keeping correct broadcasting behavior of the gradient. Will have to see if I can make this work somehow with what's there already.

**Comment by kerygma87** [comment]
This appears to be an old thread, however I'm having similar issues getting started with autograd.  @dougalm: the code you shared above throws the following error:

TypeError: Grad only applies to real scalar-output functions. Try jacobian, elementwise_grad or holomorphic_grad.

I'm using autograd 1.2, so maybe I'm missing an update?
